### PR TITLE
Support for SpeedyBee F405 v3

### PR DESCRIPTION
**Pull request by andyp1per**
Support for https://www.speedybee.com/speedybee-f405-v3-bls-50a-30x30-fc-esc-stack

Very low cost full-featured F405 FC

This is completely different to the speedybee f4 - essentially a different board

### CODE REVIEW

**Comment by andyp1per** [comment]
If anyone wants to try here are the bootloader and firmware images for copter:

[speedbeef4v3.zip](https://github.com/ArduPilot/ardupilot/files/9891539/speedbeef4v3.zip)

**Comment by rmackay9** [comment]
@Hwurzburg has volunteered will review this

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ - PWM 5-8 in group2
+ - PWM 9 in group3
+
+Channels within the same group need to use the same output rate. If
+any channel in a group uses DShot then all channels in the group need
+to use DShot. Channels 1-4 support bi-directional DShot.
+
+## Battery Monitoring
+
+The board has a builtin voltage and current sensor. The current
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md:74]:
```suggestion
The board has a internal voltage sensor and connections on the ESC connector for an external current sensor input.
```

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+sensor can read up to 130 Amps. The voltage sensor can handle up to 6S
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md:75]:
```suggestion
The voltage sensor can handle up to 6S.
```

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+LiPo batteries.
+
+The correct battery setting parameters are:
+
+ - BATT_MONITOR 4
+ - BATT_VOLT_PIN 10
+ - BATT_CURR_PIN 11
+ - BATT_VOLT_MULT 11.2
+ - BATT_AMP_PERVLT 52.7
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md:84]:
```suggestion
 - BATT_AMP_PERVLT 52.7 (will need to be adjusted for whichever current sensor is attached)
```

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+any channel in a group uses DShot then all channels in the group need
+to use DShot. Channels 1-4 support bi-directional DShot.
+
+## Battery Monitoring
+
+The board has a builtin voltage and current sensor. The current
+sensor can read up to 130 Amps. The voltage sensor can handle up to 6S
+LiPo batteries.
+
+The correct battery setting parameters are:
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md:78]:
```suggestion
The default battery parameters are:
```

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/hwdef.dat [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+PA9 USART1_TX USART1
+PA10 USART1_RX USART1
+define HAL_SERIAL1_PROTOCOL SerialProtocol_DJI_FPV
+
+# USART2 (RCIN)
+PA2 USART2_TX USART2
+PA3 USART2_RX USART2
+define HAL_SERIAL2_PROTOCOL SerialProtocol_RCIN
+
+# USART3 (CAM)
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/hwdef.dat:77]: must have protocol set to NONE or FRSKY

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/hwdef.dat [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+PC10 USART3_TX USART3 NODMA
+PC11 USART3_RX USART3 NODMA
+
+# UART4 (Bluetooth)
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/hwdef.dat:81]: must have protocol set to NONE

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/hwdef.dat [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+define HAL_STORAGE_SIZE 15360
+STORAGE_FLASH_PAGE 1
+
+# reserve 16k for bootloader and 32k for flash storage
+FLASH_RESERVE_START_KB 48
+
+# one IMU
+IMU BMI270 SPI:bmi270 ROTATION_ROLL_180
+
+# probe for a DPS280, but also look for other baros
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/hwdef.dat:125]: should not have baro probe and remove all other backends for space saving as we recently did for other 1MB boards

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+ - SERIAL4 -> UART4 (connected to internal BT module)
+ - SERIAL5 -> UART5 (ESC Telemetry)
+ - SERIAL6 -> UART6 (GPS, DMA-enabled)
+
+## RC Input
+
+RC input is configured on the R2 (UART2_RX) pin. It supports all RC
+protocols except PPM. For protocols requiring half-duplex serial to transmit
+telemetry (such as FPort) you should configure SERIAL2 with half-duplex, pin-swap
+and inversion enabled.
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md:38]:
```suggestion
For CRSF/ELRS/SRXL2 connection of the receiver to T2 will also be required.
```

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+ - SERIAL3 -> UART3 (CAM)
+ - SERIAL4 -> UART4 (connected to internal BT module)
+ - SERIAL5 -> UART5 (ESC Telemetry)
+ - SERIAL6 -> UART6 (GPS, DMA-enabled)
+
+## RC Input
+
+RC input is configured on the R2 (UART2_RX) pin. It supports all RC
+protocols except PPM. For protocols requiring half-duplex serial to transmit
+telemetry (such as FPort) you should configure SERIAL2 with half-duplex, pin-swap
```
Comment (Hwurzburg) [inline libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md:37]:
```suggestion
```

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/hwdef.dat (hunk 7 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/hwdef.dat]: Fixed

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/hwdef.dat (hunk 6 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/hwdef.dat:82]: Fixed

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/hwdef.dat (hunk 5 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/hwdef.dat:77]: Fixed

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+  - SERIAL2_OPTIONS 7
+  
+## OSD Support
+
+The SpeedyBee F405 v3 supports OSD using OSD_TYPE 1 (MAX7456 driver).
+
+## VTX Support
+
+The JST-GH-6P connector supports a standard DJI HD VTX connection. Pin 1 of the connector is 9v so be careful not to connect
+this to a peripheral requiring 5v.
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md:53]: Can only do SBUS on SBUS pin

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+ - SERIAL1 -> UART1 (DJI-VTX, DMA-enabled)
+ - SERIAL2 -> UART2 (RCIN, DMA-enabled) 
+ - SERIAL3 -> UART3 (CAM)
+ - SERIAL4 -> UART4 (connected to internal BT module)
+ - SERIAL5 -> UART5 (ESC Telemetry)
+ - SERIAL6 -> UART6 (GPS, DMA-enabled)
+
+## RC Input
+
+RC input is configured on the R2 (UART2_RX) pin. It supports all RC
```
[the remark this answers is not in the document]
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md:35]: Ignored, because must use SBUS pin for SBUS

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
looks good, pending approval by @Hwurzburg

File: libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (Hwurzburg) [reply libraries/AP_HAL_ChibiOS/hwdef/speedybeef4v3/README.md:35]: still not right...
```suggestion
RC input is configured on the R2 (UART2_RX) pin for most RC unidirectioinal protocols except SBUS which should be applied at the SBUS pin. PPM is not supported.
```

**Review by Hwurzburg — CHANGES_REQUESTED** [review]

**Review by Hwurzburg — APPROVED** [review]

**Comment by rmackay9** [comment]
I couldn't cleanly merge this to the Copter-4.3 branch so I've left it out of 4.3.1-rc1 for now at least.

**Comment by rmackay9** [comment]
This is included in Copter-4.3.1-rc1

**Pull request merged**